src/event.zig

```diff
@@ -3,5 +3,167 @@
+/// and it is milliseconds since Unix Epoch. The callback function will be executed
+/// from a secondary thread when handling events. The time parameter is provided for
+/// synchronization. Don't forget about data races!
+///
+/// **context**: A pointer-as-an-int reference to a global or allocated resource
+/// the function needs to know about when called.
+/// **callback**: The function that will be handling events.
+pub const EventCallback = struct {
+    context: usize,
+    callback: fn (usize, Event, i64) void,
+
+    const Self = @This();
+
+    /// Call the callback function.
+    /// **time**: Milliseconds since Unix Epoch.
+    pub fn call(self: Self, event: Event, time: i64) void {
+        self.callback(self.context, event, time);
+    }
+};
+
+/// An Event can be either a KeyEvent, MouseEvent, or a resize signal.
+/// You can switch on an Event to get its type:
+/// ```zig
+/// switch (event) {
+///     .Key => |key_event| ...
+///     .Mouse => |mouse_event| ...
+///     .Resize => |new_size| ...
+/// }
+/// ```
+pub const Event = union(EventType) {
+    Key: KeyEvent,
+    Mouse: MouseEvent,
+    Resize: Size,
+};
+
+pub const EventType = enum {
+    Key,
+    Mouse,
+    Resize,
+};
+
+/// A KeyEvent is an event triggered by a key press or release.
+///
+/// **value**: The key that has been pressed.
+/// **state**: Whether the key event is for key up or key down.
+/// **modifier**: One or several modifier keys that were held when the key was pressed.
+/// **repeated**: Applies to non-raw mode, where keys can be held down to be repeated.
+/// When this field is true, this is a repeat of an earlier KeyEvent. KeyState is always
+/// `Down` when a key is a repeat.
+pub const KeyEvent = struct {
+    value: KeyValue,
+    state: KeyState,
+    modifier: ?Modifier,
+    repeated: bool,
+};
+
+/// The Rune field is used when a key is alphanumerical: a-zA-Z, 0-9, and other character
+/// keys present on the keyboard. Some special characters like Tab ('\t') and space (' ')
+/// will be a Rune value. F1-F12 are found in the Function field. Home, ScrollUp, ScrollDown,
+/// Backspace, Return, and more are found in the Control field.
+pub const KeyValue = union(KeyType) {
+    Rune: Rune,
+    Function: KeyFunction,
+    Control: KeyControl,
+};
+
+pub const KeyType = enum {
+    Rune,
+    Function,
+    Control,
+};
+
+/// A modifier is a key that has been held when another key is pressed or a mouse
+/// event has occurred. This enum is a bitfield, so none, one, or any number of
+/// these fields can be selected in the same enum.
+/// ```zig
+/// if (modifier & Modifier.Alt > 0 or modifier & Modifier.RightAlt > 0) {
+///     // An alt key has been pressed ...
+/// }
+/// ```
+pub const Modifier = enum(u7) {
+    Shift = 0b1,
+    Alt = 0b10,
+    RightAlt = 0b100,
+    Ctrl = 0b1000,
+    RightCtrl = 0b10000,
+    Meta = 0b100000,
+    Function = 0b1000000,
+    _
```

Comment:
```suggestion
    _,
```
Should add comma to prevent `zig fmt` from putting all `enum` fields into a single line

Reply:
OH. I ran `zig fmt` and it did that and I was so confused. Will fix thx